### PR TITLE
ISDK-2091: Improve AudioSinkExample (2.3.0-preview)

### PR DESCRIPTION
**Pull request by ceaglest**
This PR improves AudioSinkExample, making the sinks more adaptable to different channel and sample rate configurations. This work was a initially done as part of test and internal app betterments for twilio-video-ios.

1. Refactor ExampleSpeechRecognizer to be channel agnostic.
2. ExampleAudioRecorder dynamically determines formats.
* Wait for the first valid audio sample.
* Create AVAssetWriter/Input on demand with the correct channels and sample rate.
3. Mention ExampleAudioRecorder silence detection in the known issues section of README.md.

### CODE REVIEW

File: AudioSinkExample/AudioSinks/ExampleAudioRecorder.m [hunk 1]
```diff
@@ -103,6 +110,31 @@ - (void)stopRecording {
...
+    OSStatus status = CMBlockBufferGetDataPointer(blockBuffer, 0, NULL, &inputBytes, &inputSamples);
+
+    if (status != kCMBlockBufferNoErr) {
+        NSLog(@"Failed to get data pointer: %d", status);
+        return YES;
+    }
+
+    // Check for silence. This technique is not efficient, it might be better to sum the values of the vector instead.
+    BOOL silence = YES;
+    for (int i = 0; i < inputBytes; i+=2) {
```
Comment (piyushtank) [inline AudioSinkExample/AudioSinks/ExampleAudioRecorder.m]: s/`int`/`size_t`
```for (size_t i = 0; i < inputBytes; i+=2) {```

File: AudioSinkExample/AudioSinks/ExampleSpeechRecognizer.m [hunk 2]
```diff
@@ -84,30 +79,25 @@ - (void)stopRecognizing {
...
-
-    // Detect and discard the initial invalid samples...
-    // Waits for the track to start producing the expected audio channels.
-    if (basicDescription->mChannelsPerFrame != _numberOfChannels) {
-        return;
-    }
-
     AVAudioFrameCount frameCount = (AVAudioFrameCount)CMSampleBufferGetNumSamples(audioSample);
+
+    // SFSpeechAudioBufferRecognitionRequest does not handle stereo PCM inputs correctly, so always deliver mono.
```
Comment (piyushtank) [inline AudioSinkExample/AudioSinks/ExampleSpeechRecognizer.m]: 👍

File: AudioSinkExample/AudioSinks/ExampleAudioRecorder.m [hunk 3]
```diff
@@ -122,19 +154,23 @@ + (NSURL *)recordingURLWithIdentifier:(NSString *)identifier {
...
+            return;
+        } else {
+            [self startRecordingWithTimestamp:presentationTimestamp basicDescription:basicDescription];
+        }
+    } else {
+        // Sanity check on our assumptions.
+        NSAssert(basicDescription->mChannelsPerFrame == _numberOfChannels,
+                 @"Channel mismatch. was: %d now: %d", _numberOfChannels, basicDescription->mChannelsPerFrame);
+        NSAssert(basicDescription->mSampleRate == _sampleRate,
+                 @"Sample rate mismatch. was: %f now: %f", _sampleRate, basicDescription->mSampleRate);
```
[the remark this answers is not in the document]
Comment (ceaglest) [reply AudioSinkExample/AudioSinks/ExampleAudioRecorder.m]: I don't believe so. The code chosen determines the sample rate that is sent over the wire. If your bluetooth device doesn't match then its audio will be resampled on the encoder side.

File: AudioSinkExample/AudioSinks/ExampleAudioRecorder.m [hunk 4]
```diff
@@ -122,19 +154,23 @@ + (NSURL *)recordingURLWithIdentifier:(NSString *)identifier {
...
+    const AudioStreamBasicDescription *basicDescription = CMAudioFormatDescriptionGetStreamBasicDescription(formatDescription);
     CMTime presentationTimestamp = CMSampleBufferGetPresentationTimeStamp(audioSample);
 
+    // We defer recording until the first sample in order to determine the appropriate channel layout and sample rate.
     if (CMTIME_IS_INVALID(self.recorderTimestamp)) {
-        NSLog(@"Received first valid sample. Starting recording session.");
-        [self.audioRecorder startSessionAtSourceTime:presentationTimestamp];
-        self.recorderTimestamp = presentationTimestamp;
+        // Detect and discard initial 16 kHz silence, before the first real samples are received from a remote source.
+        if (basicDescription->mSampleRate == 16000. && [self detectSilence:audioSample]) {
```
[the remark this answers is not in the document]
Comment (ceaglest) [reply AudioSinkExample/AudioSinks/ExampleAudioRecorder.m]: > If audio track continues to provide silent audio samples @ 16K are received, will it never start the recording?

It's possible, yes. Detecting silence is somewhat of a workaround, but WebRTC doesn't make this case distinct.

> does our sink API call stopRendering and startRendering in that case?

No. Our sink API just forwards the samples that we get from the media engine.

**Comment by ceaglest** [comment]
@piyushtank I responded to your comments and addressed what I could in source. We can discuss live to wrap up if you want.

**Pull request merged**